src/Simulation/qdk_sim_rs/src/python.rs

```diff
@@ -0,0 +1,243 @@
+            "features": built_info::FEATURES,
+            "target": built_info::TARGET
+        });
+        serde_json::to_string(&build_info).unwrap()
+    }
+
+    #[pyfn(m, "pauli_test")]
+    fn pauli_test(_py: Python, p: Pauli) -> Pauli {
+        p
+    }
+
+    m.add_class::<Tableau>()?;
+    m.add_class::<PyState>()?;
+    m.add_class::<PyProcess>()?;
+    m.add_class::<PyInstrument>()?;
+    m.add_class::<PyNoiseModel>()?;
+    Ok(())
+}
+
+// Design notes:
+//
+// Some important Rust concepts, such as enums, do not forward "nicely" to
+// Python types. When that happens, we make new types here to wrap the pure-Rust
+// equivalents.
+
+#[pyclass(name = "State")]
+#[derive(Debug, Clone)]
+pub struct PyState {
+    data: State,
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyState {
+    fn __repr__(&self) -> String {
+        format!("<State {:?}>", self.data)
+    }
+}
+
+#[pymethods]
+impl PyState {
+    #[staticmethod]
+    pub fn new_mixed(n_qubits: usize) -> Self {
+        Self {
+            data: State::new_mixed(n_qubits),
+        }
+    }
+
+    #[staticmethod]
+    pub fn new_stabilizer(n_qubits: usize) -> Self {
+        Self {
+            data: State::new_stabilizer(n_qubits),
+        }
+    }
+
+    pub fn as_json(&self) -> String {
+        self.data.as_json()
+    }
+}
+
+#[pyclass(name = "Process")]
+#[derive(Debug)]
+pub struct PyProcess {
+    data: Process,
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyProcess {
+    fn __repr__(&self) -> String {
+        format!("<Process {:?}>", self.data)
+    }
+}
+
+#[pymethods]
+impl PyProcess {
+    #[staticmethod]
+    pub fn new_pauli_channel(data: Vec<(f64, Vec<Pauli>)>) -> Self {
+        PyProcess {
+            data: Process::new_pauli_channel(data),
+        }
+    }
+
+    pub fn as_json(&self) -> String {
+        self.data.as_json()
+    }
+
+    pub fn apply(&self, state: PyState) -> PyResult<PyState> {
+        let data = self
+            .data
+            .apply(&state.data)
+            .map_err(|e| PyErr::new::<PyRuntimeError, String>(e))?;
+        Ok(PyState { data })
+    }
+
+    pub fn apply_to(&self, idx_qubits: Vec<usize>, state: PyState) -> PyResult<PyState> {
+        let data = self
+            .data
+            .apply_to(idx_qubits.as_slice(), &state.data)
+            .map_err(|e| PyErr::new::<PyRuntimeError, String>(e))?;
+        Ok(PyState { data })
+    }
+}
+
+#[pyclass(name = "Instrument")]
+#[derive(Debug)]
+pub struct PyInstrument {
+    data: Instrument,
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyInstrument {
+    fn __repr__(&self) -> String {
+        format!("<Instrument {:?}>", self.data)
+    }
+}
+
+#[pymethods]
+impl PyInstrument {
+    pub fn sample(&self, idx_qubits: Vec<usize>, state: &PyState) -> (usize, PyState) {
+        let (result, data) = self.data.sample(idx_qubits.as_slice(), &state.data);
+        (result, PyState { data })
+    }
+
+    #[staticmethod]
+    #[args(pr_readout_error = "0.0")]
```

Comment:
Is this line setting a default value for the argument? If so, does that work for Rust callers or is that purely a Python interop feature/attribute?

Reply:
This is purely a Python interop feature; Python functions allow declaring keyword arguments with default values, such that this is a nice convenience to allow Python users to call `Instrument.new_z_measurement()` without specifying a readout error probability.